File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/homes/HomePrivacyDtos.kt
package app.pantopus.android.data.api.models.homes

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * P3F / A14.2 — DTOs for the per-home privacy toggle set.
 * Route: `backend/routes/homePrivacy.js`.
 */

/** Envelope for `GET/PATCH /api/homes/:id/privacy` — `{ privacy }`. */
@JsonClass(generateAdapter = true)
data class HomePrivacyResponse(
    val privacy: HomePrivacyDto,
)

/** The 9 per-home privacy toggles + home id (snake_case on the wire). */
@JsonClass(generateAdapter = true)
data class HomePrivacyDto(
    @Json(name = "home_id") val homeId: String,
    @Json(name = "guest_approval") val guestApproval: Boolean,
    @Json(name = "member_name_visibility") val memberNameVisibility: Boolean,
    @Json(name = "address_precision") val addressPrecision: Boolean,
    @Json(name = "activity_visibility") val activityVisibility: Boolean,
    @Json(name = "map_opt_out") val mapOptOut: Boolean,
    @Json(name = "notification_previews") val notificationPreviews: Boolean,
    @Json(name = "doc_lock") val docLock: Boolean,
    @Json(name = "photo_blur") val photoBlur: Boolean,
    @Json(name = "vault_auto_lock") val vaultAutoLock: Boolean,
)

/**
 * Body for `PATCH /api/homes/:id/privacy`. All fields optional — Moshi omits
 * nulls by default, so only the flipped toggle is sent.
 */
@JsonClass(generateAdapter = true)
data class UpdateHomePrivacyRequest(
    @Json(name = "guest_approval") val guestApproval: Boolean? = null,
    @Json(name = "member_name_visibility") val memberNameVisibility: Boolean? = null,
    @Json(name = "address_precision") val addressPrecision: Boolean? = null,
    @Json(name = "activity_visibility") val activityVisibility: Boolean? = null,
    @Json(name = "map_opt_out") val mapOptOut: Boolean? = null,
    @Json(name = "notification_previews") val notificationPreviews: Boolean? = null,
    @Json(name = "doc_lock") val docLock: Boolean? = null,
    @Json(name = "photo_blur") val photoBlur: Boolean? = null,
    @Json(name = "vault_auto_lock") val vaultAutoLock: Boolean? = null,
)

File: frontend/apps/ios/Pantopus/Core/Networking/Models/Homes/HomePrivacyDTOs.swift
//
//  HomePrivacyDTOs.swift
//  Pantopus
//
//  P3F / A14.2 — DTOs for the per-home privacy toggle set.
//  Route: `backend/routes/homePrivacy.js`.
//

import Foundation

/// `GET/PATCH /api/homes/:id/privacy` envelope — `{ privacy }`.
public struct HomePrivacyResponse: Decodable, Sendable, Hashable {
    public let privacy: HomePrivacyDTO
}

/// The 9 per-home privacy toggles + home id. Column names are the snake_case
/// of the camelCase toggle ids the `HomeSecurityViewModel` keys on.
public struct HomePrivacyDTO: Decodable, Sendable, Hashable {
    public let homeId: String
    public let guestApproval: Bool
    public let memberNameVisibility: Bool
    public let addressPrecision: Bool
    public let activityVisibility: Bool
    public let mapOptOut: Bool
    public let notificationPreviews: Bool
    public let docLock: Bool
    public let photoBlur: Bool
    public let vaultAutoLock: Bool

    private enum CodingKeys: String, CodingKey {
        case homeId = "home_id"
        case guestApproval = "guest_approval"
        case memberNameVisibility = "member_name_visibility"
        case addressPrecision = "address_precision"
        case activityVisibility = "activity_visibility"
        case mapOptOut = "map_opt_out"
        case notificationPreviews = "notification_previews"
        case docLock = "doc_lock"
        case photoBlur = "photo_blur"
        case vaultAutoLock = "vault_auto_lock"
    }

    /// Project into the `[rowId: Bool]` map the view-model keys on. Row ids
    /// are the camelCase toggle ids (`HomeSecurityViewModel.Toggles`).
    public var toggles: [String: Bool] {
        [
            "guestApproval": guestApproval,
            "memberNameVisibility": memberNameVisibility,
            "addressPrecision": addressPrecision,
            "activityVisibility": activityVisibility,
            "mapOptOut": mapOptOut,
            "notificationPreviews": notificationPreviews,
            "docLock": docLock,
            "photoBlur": photoBlur,
            "vaultAutoLock": vaultAutoLock
        ]
    }
}

/// PATCH body — only the changed toggle(s) are encoded (the camelCase row id
/// maps to its snake_case column). Mirrors `UpdateHomeTaskRequest`'s
/// omit-nil-on-the-wire pattern.
public struct UpdateHomePrivacyRequest: Encodable, Sendable {
    public let toggles: [String: Bool]

    public init(toggles: [String: Bool]) {
        self.toggles = toggles
    }

    private enum CodingKeys: String, CodingKey {
        case guestApproval = "guest_approval"
        case memberNameVisibility = "member_name_visibility"
        case addressPrecision = "address_precision"
        case activityVisibility = "activity_visibility"
        case mapOptOut = "map_opt_out"
        case notificationPreviews = "notification_previews"
        case docLock = "doc_lock"
        case photoBlur = "photo_blur"
        case vaultAutoLock = "vault_auto_lock"
    }

    private static let keyForRowId: [String: CodingKeys] = [
        "guestApproval": .guestApproval,
        "memberNameVisibility": .memberNameVisibility,
        "addressPrecision": .addressPrecision,
        "activityVisibility": .activityVisibility,
        "mapOptOut": .mapOptOut,
        "notificationPreviews": .notificationPreviews,
        "docLock": .docLock,
        "photoBlur": .photoBlur,
        "vaultAutoLock": .vaultAutoLock
    ]

    public func encode(to encoder: any Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        for (rowId, value) in toggles {
            if let key = Self.keyForRowId[rowId] {
                try container.encode(value, forKey: key)
            }
        }
    }
}
